src/aiida_quantumespresso/parsers/base.py

```diff
@@ -25,39 +88,127 @@ def emit_logs(self, logging_dictionaries, ignore=None):
-                    except AttributeError:
-                        pass
+                    getattr(self.logger, level)(stripped)
 
-    def exit(self, exit_code):
-        """Log the exit message of the give exit code with level `ERROR` and return the exit code.
+    def check_base_errors(self, logs: AttributeDict) -> Optional[ExitCode]:
+        """Check the ``logs`` for the following "basic" parsing error and return a (formatted) version:
 
-        This is a utility function if one wants to return from the parse method and automically add the exit message
-        associated to the exit code as a log message to the node: e.g. `return self.exit(self.exit_codes.LABEL))`
+        * ``ERROR_OUTPUT_STDOUT_MISSING``
+        * ``ERROR_OUTPUT_STDOUT_READ``
+        * ``ERROR_OUTPUT_STDOUT_PARSE``
 
-        :param exit_code: an `ExitCode`
-        :return: the exit code
+        These errors mean that there is no ``stdout`` to parse.
+
+        The ``ERROR_OUTPUT_STDOUT_INCOMPLETE`` error is not checked here because in this case there might still be
+        useful information in the ``stdout``.
+        """
+
+        for exit_code in [
+            'ERROR_OUTPUT_STDOUT_MISSING',
+        ]:
+            if exit_code in logs.error:
+                return self.exit_codes.get(exit_code)
+
+        # These exit codes have additional information that needs to be formatted in the message.
+        for exit_code in [
+            'ERROR_OUTPUT_STDOUT_READ',
+            'ERROR_OUTPUT_STDOUT_PARSE'
+        ]:
+            if exit_code in logs.error:
+                exception = logs.error[logs.index(exit_code) + 1]
+                return self.exit_codes.get(exit_code).format(exception=exception)
+
+    def exit(self, exit_code: ExitCode = None, logs: AttributeDict = None) -> ExitCode:
+        """Log all messages in the ``logs`` as well as the ``exit_code`` message and return the correct exit code.
+
+        This is a utility function if one wants to return from the parse method and automically add the ``logs`` and
+        exit message associated to and exit code as a log message to the node: e.g.
+        ``return self._exit(self.exit_codes.LABEL))``
+
+        If no ``exit_code`` is provided, the method will check if an ``exit_status`` has already been set on the node
+        and return the corresponding ``ExitCode`` in this case. If not, ``ExitCode(0)`` is returned.
+
+        :param logs: log dictionaries
+        :param exit_code: an ``ExitCode``
+        :return: The correct exit code
         """
-        self.logger.error(exit_code.message)
+        if logs:
+            self.emit_logs(logs)
+
+        if exit_code is not None:
+            self.logger.error(exit_code.message)
+        elif self.node.exit_status is not None:
+            exit_code = ExitCode(self.node.exit_status, self.node.exit_message)
+        else:
+            exit_code = ExitCode(0)
+
         return exit_code
```

Comment:
This logic is not correct. Imagine the following situation: job gets submitted, but the node crashes. This is parsed by the `SlurmScheduler` and it sets `140:ERROR_NODE_FAILURE` on the `CalcJobNode`. Since the job crashed, `stdout` will either be missing or probably incomplete and so the base output parser will return the `ERROR_STDOUT_MISSING` exit code (or comparable) and in this logic, this will therefore _always_ override the scheduler status. But I would argue that the scheduler exit code is more useful in this case and we definitely want to keep it.

I think ultimately, each concrete parser implementation needs to parse the output and check _itself_ whether a scheduler exit code already exists and determine whether it can provide a more specific/informative exit code.

Reply:
This currently not causing the tests to fail, because there only is such a test for `PwParser` and there you haven't made use of this method since you wanted to not touch it. But this also means that we probably won't be able to use it. Maybe we can leave this for now then as it is, but we will have to anticipate changing it in the future when other parsers will want to become aware of scheduler exit codes and not override them

Reply:
Good points. I think the logic was indeed that in case the developer wants to return a scheduler exit code they should look for and return it in the `parse` method. This one just checks that in case the developer doesn't want to do this for whatever reason, the exit code is still added in case `self.exit()` is called at the end.

But I suppose scheduler failures will almost _always_ be more informative/useful?